### PR TITLE
Build: iOS: Use Qt6 and bump build environment

### PR DESCRIPTION
**Pull request by ann0see**
Adds a Crash-fix by @danryu for a new ios Qt 6 build. Unfortunately, Qt6 makes the app basically unusable on small devices. This should be fixed after this PR got in:
Qt6 breaks iOS connect dialog

**Short description of changes**

Updates iOS to use Qt6 (coded by @hoffie)

CHANGELOG: Update iOS build to use Qt6. For now, only larger devices are usable. Please compile with Qt5 if you use small devices.

**Context: Fixes an issue?**

Fixes: #2711
Fixes: #2939

**Does this change need documentation? What needs to be documented and how?**

**Status of this Pull Request**
Ready for review

**What is missing until this pull request can be merged?**

Nothing. However, a follow up for the GUI fix (?) should come in after the merge of this.

## Checklist

-  [x] I've verified that this Pull Request follows the [general code principles](https://github.com/jamulussoftware/jamulus/blob/main/CONTRIBUTING.md#jamulus-projectsource-code-general-principles)
-  [x] I tested my code and it does what I want
-  [x] My code follows the [style guide](https://github.com/jamulussoftware/jamulus/blob/main/CONTRIBUTING.md#source-code-consistency) 
-  [x] I waited some time after this Pull Request was opened and all GitHub checks completed without errors. 
-  [x] I've filled all the content above

### CODE REVIEW

**Comment by ann0see** [comment]
@ngocdh as you initially fixed the iOS GUI (I think), do you have any insights on what could be the issue with Qt 6?

**Comment by ngocdh** [comment]
> @ngocdh as you initially fixed the iOS GUI (I think), do you have any insights on what could be the issue with Qt 6?

Hi @ann0see I only worked on iOS audio code and maybe some UI related issues but not iOS related if I recall correctly. I did try to fix some mobile UI issues, but not knowing Qt, I made it even worse so eventually gave up.

**Comment by ann0see** [comment]
Ok.  Then someone else fixed them.

**Comment by ngocdh** [comment]
#1450 maybe?

**Comment by ann0see** [comment]
Ah yes. Maybe something is not set correctly in the conditional or we have the multiple window support enabled.

File: src/clientsettingsdlg.cpp [hunk 1]
```diff
@@ -53,6 +53,11 @@ CClientSettingsDlg::CClientSettingsDlg ( CClient* pNCliP, CClientSettings* pNSet
     layout()->setMenuBar ( pMenu );
 #endif
```
Comment (ann0see) [inline src/clientsettingsdlg.cpp:54]: Smells like avoidable duplication.

**Comment by ann0see** [comment]
Ok. I think the chat dialog needs to be fixed still - but that was the correct direction...

File: src/connectdlg.cpp [hunk 2]
```diff
@@ -148,14 +148,14 @@ CConnectDlg::CConnectDlg ( CClientSettings* pNSetP, const bool bNewShowCompleteR
...
-    TimerInitialSort.setSingleShot ( true ); // only once after list request
-
-#if defined ( Q_OS_ANDROID ) || defined ( ANDROID ) || defined ( Q_OS_IOS )
-    // for the android/iOS version maximize the window
+#if defined( Q_OS_ANDROID ) || defined( ANDROID ) || defined( Q_OS_IOS )
+    // for the Android/iOS version maximize the window
     setWindowState ( Qt::WindowMaximized );
 #endif
 
+    // setup timers
```
Comment (ann0see) [inline src/connectdlg.cpp:156]: I'm not sure if the order we set the maximized state matters.

File: src/connectdlg.cpp [hunk 3]
```diff
@@ -148,14 +148,14 @@ CConnectDlg::CConnectDlg ( CClientSettings* pNSetP, const bool bNewShowCompleteR
     // set a placeholder text to explain how to filter occupied servers (#397)
     edtFilter->setPlaceholderText ( tr ( "Filter text, or # for occupied servers" ) );
 
-    // setup timers
```
Comment (ann0see) [inline src/connectdlg.cpp:151]: I'm not sure if the order we set the maximized state matters. At least in the version which kind of worked - I first needed to rotate the screen to allow it getting the correct sizing - but that's way better than it was.

**Comment by ann0see** [comment]
Ok. At least connecting works servers with welcome message bring issues due to the crash if close is tapped (this doesn't happen if I'm not connected to a server but the chat window is open)

File: src/clientsettingsdlg.cpp (hunk 1 above)
Comment (pljones) [reply src/clientsettingsdlg.cpp:54]: Do we live with this for now as it's unrelated?

Comment (ann0see) [reply src/clientsettingsdlg.cpp:54]: Unrelated, but worth an issue

**Comment by ann0see** [comment]
The iOS log says at the start of the iOS app

```
May 21 20:22:02 Jamulus[58448] <Notice>: Detected system locale encoding (US-ASCII, locale "C") is not UTF-8.
Qt shall use a UTF-8 locale ("UTF-8") instead. If this causes problems,
reconfigure your locale. See the locale(1) manual for more information.
```
I don't think that's related but interesting to know. I'll downgrade the SDK. Maybe that fixes it? I don't really understand the issue. To me it seems like we dereference a NULL pointer.

```
May 21 20:22:25 SpringBoard(KeyboardArbiter)[55730] <Error>: [<_UIKeyboardArbiter: 0x28236ed80; <<_UIKeyboardArbiter_ForSpringBoard: 0x281653c90>; scene = <FBSScene: 0x28368ba80; identifier: com.apple.UIKit.remote-keyboard; valid: YES> {
    settings = <UIApplicationSceneSettings: 0x283602280> {
        displayConfiguration = "<FBSDisplayConfiguration: 0x28216f560; Main; mode: "320x568@2x 60Hz sRGB SDR">";
        frame = {{0, 0}, {320, 568}};
        level = 2.0;
        interfaceOrientation = unknown (0);
        foreground = NO;
        interruptionPolicy = default;
        isOccluded = BSSettingFlagNotSet;
        activityMode = default;
        jetsamMode = default;
    };
    clientSettings = <UIApplicationSceneClientSettings: 0x2803572c0> {
        layers = (empty);
        preferredLevel = 2.0;
        preferredInterfaceOrientation = unknown (0);
    };
    hostProcess = <FBApplicationProcess: 0x10310b770; daemon<com.apple.SpringBoard>:55730(v286BE); taskState: Running; visibility: Unknown>;
}>;
disabler = (null);
lastEventSource = 0;
current state = (null);
active handle = (null);
handles = (
    "<_UIKeyboardArbiterHandle: 0x282f72cb0; PID 55730: com.apple.springboard <com.apple.frontboard.systemappservices::com.apple.springboard>; hosting PIDs {(\134n)}; level 2.000000; active NO [wants NO]; suppression 0; iav 0.000000; on screen NO>",
    "<_UIKeyboardArbiterHandle: 0x282f41b80; PID 56371: (null) <(null)>; hosting PIDs {(\134n)}; level 0.000000; active NO [wants NO]; suppression 0; iav 0.000000; on screen NO>",
    "<_UIKeyboardArbiterHandle: 0x282f4c9a0; PID 57286: com.apple.MailCompositionService <(null)> (suspended); hosting PIDs {(\134n)}; level 0.000000; active NO [wants NO]; suppression 0; iav 0.000000; on screen NO>",
    "<_UIKeyboardArbiterHandle: 0x282f27700; PID 57909: com.apple.PassbookUIService <(null)> (suspended); hosting PIDs {(\134n)}; level 0.000000; active NO [wants NO]; suppression 0; iav 0.000000; on screen NO>"
)>] No handle found for currently focused PID: 58448; sceneIdentity: com.apple.frontboard.systemappservices::sceneID%3Aio.jamulus.Jamulus.<redacted>-default
```
Is probably the place where the error occurs.
```
May 21 20:22:26 ReportCrash(CoreAnalytics)[58352] <Notice>: Sending event: com.apple.stability.crash {"bundleID":"io.jamulus.Jamulus.<redacted>","bundleVersion":"1","exceptionCodes":"0x0000000000000001, 0x0000000000000000(\134n    1,\134n    0\134n)EXC_BAD_ACCESSSIGSEGVKERN_INVALID_ADDRESS at 0x0000000000000000","incidentID":"DFD5F2F2-D1FF-4E87-9C72-BDC2882CCEB8","logwritten":1,"process":"Jamulus","terminationReasonExceptionCode":"0xb","terminationReasonNamespace":"SIGNAL"}
May 21 20:22:26 analyticsd[100] <Notice>: Received event: com.apple.stability.crash {"bundleID":"io.jamulus.Jamulus.<redacted>","bundleVersion":"1","exceptionCodes":"0x0000000000000001, 0x0000000000000000(\134n    1,\134n    0\134n)EXC_BAD_ACCESSSIGSEGVKERN_INVALID_ADDRESS at 0x0000000000000000","incidentID":"DFD5F2F2-D1FF-4E87-9C72-BDC2882CCEB8","logwritten":1,"process":"Jamulus","terminationReasonExceptionCode":"0xb","terminationReasonNamespace":"SIGNAL"}
May 21 20:22:26 analyticsd[100] <Notice>: Aggregated. Transform: StabilityCrashNumerator3WithBundleVersion Dirty: 1 Event: com.apple.stability.crash {"bundleID":"io.jamulus.Jamulus.<redacted>","bundleVersion":"1","exceptionCodes":"0x0000000000000001, 0x0000000000000000(\134n    1,\134n    0\134n)EXC_BAD_ACCESSSIGSEGVKERN_INVALID_ADDRESS at 0x0000000000000000","incidentID":"DFD5F2F2-D1FF-4E87-9C72-BDC2882CCEB8","logwritten":1,"process":"Jamulus","terminationReasonExceptionCode":"0xb","terminationReasonNamespace":"SIGNAL","timestamp":1684693346614102}
May 21 20:22:26 analyticsd[100] <Notice>: Aggregated. Transform: StabilityCrashNumerator3WithIncidentID Dirty: 1 Event: com.apple.stability.crash {"bundleID":"io.jamulus.Jamulus.<redacted>","bundleVersion":"1","exceptionCodes":"0x0000000000000001, 0x0000000000000000(\134n    1,\134n    0\134n)EXC_BAD_ACCESSSIGSEGVKERN_INVALID_ADDRESS at 0x0000000000000000","incidentID":"DFD5F2F2-D1FF-4E87-9C72-BDC2882CCEB8","logwritten":1,"process":"Jamulus","terminationReasonExceptionCode":"0xb","terminationReasonNamespace":"SIGNAL","timestamp":1684693346614102}
May 21 20:22:26 analyticsd[100] <Notice>: Aggregated. Transform: StabilityCrashNumerator3 Dirty: 1 Event: com.apple.stability.crash {"bundleID":"io.jamulus.Jamulus.<redacted>","bundleVersion":"1","exceptionCodes":"0x0000000000000001, 0x0000000000000000(\134n    1,\134n    0\134n)EXC_BAD_ACCESSSIGSEGVKERN_INVALID_ADDRESS at 0x0000000000000000","incidentID":"DFD5F2F2-D1FF-4E87-9C72-BDC2882CCEB8","logwritten":1,"process":"Jamulus","terminationReasonExceptionCode":"0xb","terminationReasonNamespace":"SIGNAL","timestamp":1684693346614102}
```
seems to be where the system logs the error.

**Comment by ann0see** [comment]
Now a linking error... I'll try to downgrade to Qt5 and then raise another PR if that works.

**Comment by ann0see** [comment]
Ok. It could be that just adding the flags doesn't solve the problem. We might need to load some plugins manually

https://stackoverflow.com/questions/45508043/qt-ios-linker-error-entry-point-main-undefined

**Comment by ann0see** [comment]
The build seems to also produce binaries for the iOS simulator - so the IPA would also run on the x86 mac simulator - which is not what we want. I think deploy_ios.sh needs to be tweaked too...

**Comment by ann0see** [comment]
I believe the change in deploy_ios.sh removed the simulator as target. This needs further investigation (might be related to the issue not needing the macOS workaround). 

I believe this PR should be considered as in development now. The Qt5 one should be ready.

**Comment by ann0see** [comment]
The build failure sounds bad here. Xcode would like us to use `"` instead of `<` for includes as far as I understand? But doesn't `<` have a meaning that it's in the include path and something standard while `"` means that it's a header file of the project?

**Comment by pljones** [comment]
> Use quote-style include (`"example.h"`) for project headers, and reserve angle-bracket include (`<example.h>`) for system headers.

So no change there.  The only thing C++ seems to do is drop the `.h`?  (So `<vector>` rather than `#include <fcntl.h>`.)

File: ios/deploy_ios.sh [hunk 4]
```diff
@@ -1,10 +1,31 @@
...
+            exit 1
+            ;;
+    esac
+done
+
 ## Builds an ipa file for iOS. Should be run from the repo-root
 
 # Create Xcode file and build
-qmake -spec macx-xcode Jamulus.pro
+eval "${qmake_path} Jamulus.pro"
```
Comment (ann0see) [inline ios/deploy_ios.sh:28]: I don't like eval here

**Comment by ann0see** [comment]
Ok. Could get another run on the simulator and got a crash:

```
2023-06-06 21:33:10.326150+0200 Jamulus[4974:50829] [SceneConfiguration] Info.plist contained no UIScene configuration dictionary (looking for configuration named "(no name)")
2023-06-06 21:33:10.398742+0200 Jamulus[4974:50829] - allocated port number: 22134
2023-06-06 21:33:10.977508+0200 Jamulus[4974:50829] [plugin] AddInstanceForFactory: No factory registered for id <CFUUID 0x600003090120> F8BB1C28-BAE8-11D6-9C31-00039315CD46
2023-06-06 21:33:32.011739+0200 Jamulus[4974:50829] fopen failed for data file: errno = 2 (No such file or directory)
2023-06-06 21:33:32.012905+0200 Jamulus[4974:50829] Errors found! Invalidating cache...
2023-06-06 21:33:32.929661+0200 Jamulus[4974:50829] fopen failed for data file: errno = 2 (No such file or directory)
2023-06-06 21:33:32.929946+0200 Jamulus[4974:50829] Errors found! Invalidating cache...
2023-06-06 21:33:44.013115+0200 Jamulus[4974:50829] QHBoxLayout(0x6000030c17e0, name = "horizontalLayout_7") is laying out QTabWidget(0x600003ebc0f0, name="tabSettings") outside of the contents rect of CClientSettingsDlg(0x7ff7ba208398, name="CClientSettingsDlgBase")
2023-06-06 21:33:44.013782+0200 Jamulus[4974:50829] QHBoxLayout(0x6000030c17e0, name = "horizontalLayout_7") is laying out QTabWidget(0x600003ebc0f0, name="tabSettings") outside of the contents rect of CClientSettingsDlg(0x7ff7ba208398, name="CClientSettingsDlgBase")
2023-06-06 21:33:44.015753+0200 Jamulus[4974:50829] QHBoxLayout(0x6000030c17e0, name = "horizontalLayout_7") is laying out QTabWidget(0x600003ebc0f0, name="tabSettings") outside of the contents rect of CClientSettingsDlg(0x7ff7ba208398, name="CClientSettingsDlgBase")
2023-06-06 21:33:44.016987+0200 Jamulus[4974:50829] QHBoxLayout(0x6000030c17e0, name = "horizontalLayout_7") is laying out QTabWidget(0x600003ebc0f0, name="tabSettings") outside of the contents rect of CClientSettingsDlg(0x7ff7ba208398, name="CClientSettingsDlgBase")
2023-06-06 21:33:44.018262+0200 Jamulus[4974:50829] QHBoxLayout(0x6000030c17e0, name = "horizontalLayout_7") is laying out QTabWidget(0x600003ebc0f0, name="tabSettings") outside of the contents rect of CClientSettingsDlg(0x7ff7ba208398, name="CClientSettingsDlgBase")
2023-06-06 21:33:44.018982+0200 Jamulus[4974:50829] QHBoxLayout(0x6000030c17e0, name = "horizontalLayout_7") is laying out QTabWidget(0x600003ebc0f0, name="tabSettings") outside of the contents rect of CClientSettingsDlg(0x7ff7ba208398, name="CClientSettingsDlgBase")
2023-06-06 21:33:44.023379+0200 Jamulus[4974:50829] QHBoxLayout(0x6000030c17e0, name = "horizontalLayout_7") is laying out QTabWidget(0x600003ebc0f0, name="tabSettings") outside of the contents rect of CClientSettingsDlg(0x7ff7ba208398, name="CClientSettingsDlgBase")
2023-06-06 21:33:57.247931+0200 Jamulus[4974:50829] ASSERT failure in QCoreApplication::sendEvent: "Unexpected null receiver", file /Users/qt/work/qt/qtbase/src/corelib/kernel/qcoreapplication.cpp, line 1527
CoreSimulator 857.14 - Device: iPhone 14 Pro Max (A0A2608F-FCF3-4B1B-9E96-A64EC82AD958) - Runtime: iOS 16.2 (20C52) - DeviceType: iPhone 14 Pro Max
(lldb) 
```

**Comment by ann0see** [comment]
It could also be a race condition. But that would be out of scope of Jamulus and a Qt bug?

**Comment by ann0see** [comment]
hide() instead of close() works. So tat would be an option.

**Comment by pljones** [comment]
We've already got issues on Android where the application gets hidden without going through close and not saving settings.

I'd beware of hiding rather than closing when the user asks to exit until we're catching all OS signals indicating we should save state (there's a Qt signal, IIRC).

**Comment by pljones** [comment]
Any progress on this or can it be dropped from 3.10.0 (with #2711)?

**Comment by ann0see** [comment]
I think this can be dropped. I think it could also be a Qt bug which we can't really fix.

**Comment by ann0see** [comment]
Unfortunately I've lost track on this. Qt 6 UI bugs also show up on bigger devices but that's the main issue.

**Comment by pljones** [comment]
Is there any way to split the Jamulus application changes out from the build tool chain changes?  Can one be made without the other or must they go in together?

**Comment by ann0see** [comment]
Potentially yes. As I said, This needs to be recreated anyway.

**Pull request closed without merging**